[docs] Add more details about Python formatting

jdenny-ornl opened the pull request:
Describe the darker utility.  Make it clear that black/darker's default rules should be used.

See ongoing discussion at
<https://discourse.llvm.org/t/rfc-document-and-standardize-python-code-style/68257>.

jdenny-ornl committed:
[docs] Add more details about Python formatting

Describe the darker utility.  Make it clear that black/darker's
default rules should be used.

See ongoing discussion at
<https://discourse.llvm.org/t/rfc-document-and-standardize-python-code-style/68257>.

cor3ntin commented on llvm/docs/CodingStandards.rst:
```diff
@@ -115,10 +115,19 @@ available in this version of Python.
+23.x in LLVM.
+
+When contributing a patch unrelated to formatting, use the `darker
+<https://pypi.org/project/darker/>`_ utility with its default rules to format
+any changed Python code. Doing so should ensure the patch will pass the Python
```
I do not have a specific suggestion, but could we make it clearer that `darker` is something that calls `black` rather than a different formatter?

tru replied:
Hah, yeah I was just going to post the same suggestion.

jdenny-ornl replied:
Good idea.  Done.

jdenny-ornl committed:
Clarify relationship among PEP 8, black, and darker

Also add some quick examples.

tru commented on llvm/docs/CodingStandards.rst:
```diff
@@ -113,21 +113,32 @@ section. Python code in the LLVM repository should only use language features
+
+.. code-block:: bash
+
+    $ pip install black darker
```
I think we should pin versions here, maybe use a requirements.txt or something? We have it stated above that we recommend version 23.x for now, and the point of this is that black only guarantees a stable format within the major versions.

jdenny-ornl replied:
To keep the example simple, does the following seem reasonable?

```
$ pip install black=='23.*' darker
```

tru commented on llvm/docs/CodingStandards.rst:
```diff
@@ -113,21 +113,32 @@ section. Python code in the LLVM repository should only use language features
+
+When contributing a patch unrelated to formatting, you should format only the
+Python code that the patch modifies. For this purpose, use the `darker
+<https://pypi.org/project/darker/>`_ utility to call black with its default
```
I would probably say something like "darker utility which runs default black rules over only the changed patch"..

jdenny-ornl replied:
Done.  Let me know if the wording I chose doesn't satisfy the concern.

jdenny-ornl committed:
More tweaks

Wordsmith.

Specify black version in example `pip install` command.

In the example, use `HEAD` instead of `@` as more git users probably
know the former.

In the example, drop `..:WORKTREE:` as it's hard to remember and,
fortunately, redundant.  Leave that part to the darker documentation.

tru reviewed (approved):
LGTM

DavidSpickett commented on llvm/docs/CodingStandards.rst:
```diff
@@ -113,13 +113,33 @@ section. Python code in the LLVM repository should only use language features
+    $ pip install black=='23.*' darker # install black 23.x and darker
+    $ darker test.py                   # format uncommitted changes
+    $ darker -r HEAD^ test.py          # also format changes from last commit
+    $ black test.py                    # format entire file
```
I think it's worth adding a general format all .py in the last commit, like clang-format-diff:
```
darker --revision HEAD~1 $(git diff-tree --no-commit-id --name-only HEAD -r)
```
I found that without file names darker likes to search the entire tree first, then choose what's changed, which takes ages.

jdenny-ornl replied:
Good idea. That annoyed me too.

I chose a slightly simpler `git diff` command so that it's easier to remember.  Please let me know whether that works for you.

DavidSpickett replied:
Works great, thanks.

jdenny-ornl committed:
Discuss darker called on directories
